### PR TITLE
Fix #16: Knob should snap to the correct value on tapping

### LINKED ISSUE #16: Knob not "snapping" to correct value on tapping
When tapping on the slider, the knob will go to that location exactly - whereas the slider will fit to a valid value. That will leave a gap between the slider position and the knob position.

Here's a screenshot:

image

This does not happen if you slide the knob by click and drag - only when tapping somewhere on the progress bar.

This is most noticable when max value is low (try setting it to 2 or 3)

### PR DESCRIPTION
**Pull request by blasten**
cc @morethanreal  @notwaldorf

### CODE REVIEW

File: paper-slider.html [hunk 1]
```diff
@@ -315,9 +315,11 @@
     },
 
     _positionKnob: function(ratio) {
+      var s = this.$.sliderKnob.style;
```
Comment (cdata) [inline paper-slider.html]: What is `s`?

File: paper-slider.html [hunk 2]
```diff
@@ -359,21 +361,29 @@
...
-      this._setImmediateValue(this._calcStep(
-          this._calcKnobPosition(this._x / this._w)) || 0);
       var s =  this.$.sliderKnob.style;
-      s.transform = s.webkitTransform = 'translate3d(' + (this.snaps ?
-          (this._calcRatio(this.immediateValue) * this._w) - this._startx : x) + 'px, 0, 0)';
+      var translateX;
+
+      this._x = this._startx + x;
+      this._setImmediateValue(
+          this._calcStep(this._calcKnobPosition(this._x / this._w)) || 0);
```
Comment (cdata) [inline paper-slider.html]: Usages of `this._calcStep` seem to be followed with `|| 0`. Perhaps `0` should be a built-in default return value for the `_calcStep` method?

File: paper-slider.html [hunk 3]
```diff
@@ -359,21 +361,29 @@
...
+
+      translateX = (this._calcRatio(this.immediateValue) * this._w) - this._startx;
+
+      // update knob's position
+      s.transform = s.webkitTransform = 'translate3d(' + translateX + 'px, 0, 0)';
```
Comment (cdata) [inline paper-slider.html]: Why not use `this.translate3d`?

File: paper-slider.html [hunk 4]
```diff
@@ -359,21 +361,29 @@
...
       var s =  this.$.sliderKnob.style;
-      s.transform = s.webkitTransform = 'translate3d(' + (this.snaps ?
-          (this._calcRatio(this.immediateValue) * this._w) - this._startx : x) + 'px, 0, 0)';
+      var translateX;
+
+      this._x = this._startx + x;
+      this._setImmediateValue(
+          this._calcStep(this._calcKnobPosition(this._x / this._w)) || 0);
+
+      translateX = (this._calcRatio(this.immediateValue) * this._w) - this._startx;
```
Comment (cdata) [inline paper-slider.html]: This identifier is set once and then used immediately after it is set. Perhaps we don't need the identifier?

File: paper-slider.html [hunk 5]
```diff
@@ -384,13 +394,25 @@
...
+
       this._w = this.$.sliderBar.offsetWidth;
       var rect = this.$.sliderBar.getBoundingClientRect();
       var ratio = (e.detail.x - rect.left) / this._w;
+      var prevRatio = this.ratio;
+
+      this._setTransiting(true);
       this._positionKnob(ratio);
+
       this._expandJob = this.debounce(this._expandJob, this._expandKnob, 60);
```
Comment (cdata) [inline paper-slider.html]: Not related to this change, but important: the return value of `this.debounce` is always `undefined`, so `this._expandJob` is being set to `undefined` here.

**Comment by cdata** [comment]
Is there a test?

File: paper-slider.html (hunk 4 above)
Comment (blasten) [inline paper-slider.html]: I refactored this function. It still has these variables (or const in ES6)  to provide semantic.

File: paper-slider.html (hunk 3 above)
Comment (blasten) [inline paper-slider.html]: Great, I didn't know about that helper function.

File: paper-slider.html (hunk 2 above)
Comment (blasten) [inline paper-slider.html]: I will handle this case in `this._calcStep`

File: paper-slider.html (hunk 1 above)
Comment (blasten) [inline paper-slider.html]: removed.

**Comment by blasten** [comment]
@cdata I added a test. can you ptal? thanks

**Comment by cdata** [comment]
LGTM :+1:

**Pull request merged**